metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala

```diff
@@ -403,8 +371,21 @@ final class BloopServers(
 
   private def connectToLauncher(
       bloopVersion: String,
-      bloopPort: Option[Int]
+      bloopPort: Option[Int],
+      userConfiguration: UserConfiguration
   ): Future[SocketConnection] = {
+    // we should set up Java before running Bloop in order to not restart it
+    bloopJsonPath match {
```

Comment:
This is to make sure that we don't restart Bloop after setting bloop.json for the first time.

Reply:
this should only happen when the javaHome that Bloop was using was already the same as that of Metals, or as the one we set for Metals in the setting.

Reply:
This will always happen at the start, when no bloop.json was set up previously.

Reply:
So now, if

- there's no `$HOME/.bloop/bloop.json` and
- Whether there's `metals.bloopJvmProperties` defined or not

Metals will always write the global bloop setting (when connect to bloop launcher) file to `$HOME/.bloop/bloop.json` where `javaHome` will be inferred java home location by `locate-java-home`.

---

I'm a bit concerned about the following scenario:
- Metals write the global bloop file to `$HOME/.bloop/bloop.json` where `javaHome` is javaHome of currently using java (inferred by `locate-java-home`)
- User upgrade java version (and now `locate-java-home` locate the new java's javahome.
- Metals won't update the global bloop setting file, because the user setting of Metals didn't change (and `workspace/didChangeConfiguration` doesn't fire).
  - Or the user declines Metals to write bloop global file.
- While Metals uses new java, bloop still has old javaHome in the global bloop setting file, and bloop will continue using old Java.

It might be better to
- (1) every time Metals server starts, write JVM properties to bloop global json file even if there exists a file (but I'm not sure how should we handle the existing file :thinking:)
- (2) don't write java home if a user didn't specify java home explicitly via user settings.

Reply:
@tgodzik 
Oh does workspace/didChangeConfigration will  be fired (at least by VSCode) on server starts, even if the setting has changed before Metals server starts?
In that case, it's all good :+1:

 
> The only issue will be on other editors if Metals was started and the setting was later updated manually

I'm not sure, but other clients also should fire. Anyway, it would be better to implement (1) as you mentioned, if it's handy. If it's complexed, we can go as it is, at least for now

Reply:
> Oh does workspace/didChangeConfigration will be fired (at least by VSCode) on server starts, even if the setting has changed before Metals server starts?
> In that case, it's all good

Yeah, it should be sent correctly.

> I'm not sure, but other clients also should fire. Anyway, it would be better to implement (1) as you mentioned, if it's handy. If it's complexed, we can go as it is, at least for now

So another thing is that if set javaHome before starting Bloop, we would have to restart it in case it's running, which is a bit more complex and would be outside the user's influence. For now I would leave at is and maybe improve the behaviour later on?

Reply:
Coming to think of it you currently need to connect to Bloop to restart it, so we would need to invoke Bloop CLI for that (which could be a part of a different issue though) https://github.com/scalameta/metals/issues/3542